pkg/koordlet/metricsadvisor/collectors/prom/scraper.go

```diff
@@ -0,0 +1,345 @@
+		return nil, fmt.Errorf("get metrics failed, err: %w", err)
+	}
+	klog.V(6).Infof("scrape prom metrics finished by scraper %s, metrics num %d", c.Name(), len(samples))
+
+	metricSamples, err := c.Decoder.Decode(samples, node, podMetas)
+	if err != nil {
+		return nil, fmt.Errorf("decode metrics failed, err: %w", err)
+	}
+
+	klog.V(6).Infof("decode prom samples finished by scraper %s, metrics num %d", c.Name(), len(metricSamples))
+	return metricSamples, nil
+}
+
+type DecodeConfig struct {
+	TargetMetric   string
+	TargetType     MetricType
+	SourceMetric   string
+	SourceLabels   model.LabelSet
+	ValidTimeRange time.Duration
+}
+
+type Decoder interface {
+	Decode(samples []*model.Sample, node *corev1.Node, podMetas []*statesinformer.PodMeta) ([]metriccache.MetricSample, error)
+}
+
+type DecodeFn func(samples []*model.Sample, node *corev1.Node, podMetas []*statesinformer.PodMeta) ([]metriccache.MetricSample, error)
+
+type MultiDecoderImpl struct {
+	decoders []Decoder
+}
+
+func NewMultiDecoder(decoders ...Decoder) Decoder {
+	return &MultiDecoderImpl{
+		decoders: decoders,
+	}
+}
+
+func (m *MultiDecoderImpl) Decode(samples []*model.Sample, node *corev1.Node, podMetas []*statesinformer.PodMeta) ([]metriccache.MetricSample, error) {
+	// TODO: look up multiple target metrics faster
+	var allMetricSamples []metriccache.MetricSample
+	var errs []error
+	for _, decoder := range m.decoders {
+		metricSamples, err := decoder.Decode(samples, node, podMetas)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		allMetricSamples = append(allMetricSamples, metricSamples...)
+	}
+	if len(errs) > 0 {
+		return allMetricSamples, utilerrors.NewAggregate(errs)
+	}
+	return allMetricSamples, nil
+}
+
+type DecoderImpl struct {
+	decodeFn DecodeFn
+}
+
+func NewDecoder(decodeFn DecodeFn) Decoder {
+	return &DecoderImpl{
+		decodeFn: decodeFn,
+	}
+}
+
+func NewGenericDecoder(cfg *DecodeConfig) Decoder {
+	return &DecoderImpl{
+		decodeFn: GenericDecodeFn(cfg.SourceMetric, cfg.SourceLabels, cfg.TargetMetric, cfg.TargetType, cfg.ValidTimeRange),
+	}
+}
+
+func (d *DecoderImpl) Decode(samples []*model.Sample, node *corev1.Node, podMetas []*statesinformer.PodMeta) ([]metriccache.MetricSample, error) {
+	return d.decodeFn(samples, node, podMetas)
+}
+
+func GenericDecodeFn(metricName string, labels model.LabelSet, targetName string, targetType MetricType, validTimeRange time.Duration) DecodeFn {
+	return func(samples []*model.Sample, node *corev1.Node, podMetas []*statesinformer.PodMeta) ([]metriccache.MetricSample, error) {
+		timeNow := model.Now()
+		samplesMap := map[string][]*model.Sample{}
+		for _, sample := range samples {
+			if string(sample.Metric[model.MetricNameLabel]) != metricName {
+				continue
+			}
+			// in case some metrics have no timestamp
+			if sample.Timestamp.UnixNano() <= 0 {
+				sample.Timestamp = timeNow
+			}
+			if sample.Timestamp.Sub(timeNow) > validTimeRange && sample.Timestamp.Sub(timeNow) < -validTimeRange {
```

Comment:
check the meaning of sample.Timestamp, no need to check if the ts is `scrap time` but not `origin collect time`

Reply:
scrap time but not origin collect time?

Reply:
Both are possible. Though many prom exporters do not set the timestamp during collection, some exporters like the [cadvisor](https://github.com/google/cadvisor/blob/v0.48.1/metrics/prometheus_machine.go#L262) expose the metrics with collecting timestamp via `prometheus.NewMetricWithTimestamp`.